gobblin-runtime/src/main/java/gobblin/runtime/mapreduce/MRJobLauncher.java

```diff
@@ -105,60 +105,125 @@
+        jobListener.onJobCancellation(this.jobState);
+      }
     } catch (IOException ioe) {
       LOG.error("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
       throw new JobException("Failed to kill the Hadoop MR job " + this.job.getJobID(), ioe);
+    } finally {
+      cleanUpWorkingDirectory();
     }
   }
 
   @Override
-  protected void runJob(String jobName, Properties jobProps, JobState jobState, List<WorkUnit> workUnits)
-      throws Exception {
+  public void close()
+      throws IOException {
+    try {
+      super.close();
+    } finally {
+      if (!this.isCancelled && this.job != null && !this.job.isComplete()) {
+        LOG.info("Killing Hadoop MR job " + this.job.getJobID());
+        try {
+          this.job.killJob();
+        } finally {
+          cleanUpWorkingDirectory();
+        }
+      }
+    }
+  }
 
+  @Override
+  protected void runWorkUnits(List<WorkUnit> workUnits)
+      throws Exception {
     // Add job config properties that also contains all framework config properties
-    for (String name : jobProps.stringPropertyNames()) {
-      this.conf.set(name, jobProps.getProperty(name));
+    for (String name : this.jobProps.stringPropertyNames()) {
+      this.conf.set(name, this.jobProps.getProperty(name));
+    }
+
+    // Let the job and all mappers finish even if some mappers fail
+    this.conf.set("mapred.max.map.failures.percent", "100"); // For Hadoop 1.x
+    this.conf.set("mapreduce.map.failures.maxpercent", "100"); // For Hadoop 2.x
+
+    // Do not cancel delegation tokens after job has completed (HADOOP-7002)
+    this.conf.setBoolean("mapreduce.job.complete.cancel.delegation.tokens", false);
+
+    this.mrJobDir = new Path(this.jobProps.getProperty(ConfigurationKeys.MR_JOB_ROOT_DIR_KEY), this.jobName);
+    if (this.fs.exists(this.mrJobDir)) {
+      LOG.warn("Job working directory already exists for job " + this.jobName);
+      this.fs.delete(this.mrJobDir, true);
     }
 
-    Path mrJobDir = new Path(jobProps.getProperty(ConfigurationKeys.MR_JOB_ROOT_DIR_KEY), jobName);
-    if (this.fs.exists(mrJobDir)) {
-      LOG.warn("Job working directory already exists for job " + jobName);
-      this.fs.delete(mrJobDir, true);
+    // Delete any staging directories that already exist before the Hadoop MR job starts
+    JobLauncherUtils.cleanStagingData(JobLauncherUtils.flattenWorkUnits(workUnits), LOG);
+
+    try {
+      Path jobOutputPath = prepareHadoopJob(this.jobName, this.jobProps, workUnits);
+      LOG.info("Launching Hadoop MR job " + this.job.getJobName());
+      this.job.submit();
+      // Set job tracking URL to the Hadoop job tracking URL if it is not set yet
+      if (!this.jobState.contains(ConfigurationKeys.JOB_TRACKING_URL_KEY)) {
+        this.jobState.setProp(ConfigurationKeys.JOB_TRACKING_URL_KEY, this.job.getTrackingURL());
+      }
+      // Wait for the job to complete
+      this.job.waitForCompletion(true);
+
+      if (this.isCancelled) {
```

Comment:
Is this the safest approach? If there are two threads, where Thread 1 is executing `runWorkUnits()` and Thread 2 is executing `cancelJob()`, then the following scenario may be possible.

Thread 1 -> `this.job.waitForCompletion(true)`
Thread 2 -> `cancelJob()`
Thread 2 -> `cancelJob() -> this.job.killJob()` succeeds
Thread 1 -> Exits from `this.job.waitForCompletion(true)`
Thread 1 -> `if (this.isCancelled)` evaluates to false, and the `jobState` doesn't get set properly
Thread 2 -> `this.isCancelled = true`